File: src/main/java/br/com/jiratorio/service/impl/LeadTimeConfigServiceImpl.java
package br.com.jiratorio.service.impl;

import br.com.jiratorio.domain.Board;
import br.com.jiratorio.domain.LeadTimeConfig;
import br.com.jiratorio.domain.request.LeadTimeConfigRequest;
import br.com.jiratorio.domain.response.LeadTimeConfigResponse;
import br.com.jiratorio.exception.ResourceNotFound;
import br.com.jiratorio.mapper.LeadTimeConfigMapper;
import br.com.jiratorio.repository.LeadTimeConfigRepository;
import br.com.jiratorio.service.BoardService;
import br.com.jiratorio.service.LeadTimeConfigService;
import java.util.List;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Slf4j
@Service
public class LeadTimeConfigServiceImpl extends AbstractService implements LeadTimeConfigService {

    private final LeadTimeConfigRepository leadTimeConfigRepository;

    private final BoardService boardService;

    private final LeadTimeConfigMapper leadTimeConfigMapper;

    public LeadTimeConfigServiceImpl(final LeadTimeConfigRepository leadTimeConfigRepository,
                                     final BoardService boardService,
                                     final LeadTimeConfigMapper leadTimeConfigMapper) {
        this.leadTimeConfigRepository = leadTimeConfigRepository;
        this.boardService = boardService;
        this.leadTimeConfigMapper = leadTimeConfigMapper;
    }

    @Override
    @Transactional(readOnly = true)
    public List<LeadTimeConfig> findAllByBoardId(final Long boardId) {
        log.info("Method=findAllByBoardId, boardId={}", boardId);
        return leadTimeConfigRepository.findByBoardId(boardId);
    }

    @Override
    @Transactional(readOnly = true)
    public List<LeadTimeConfigResponse> findAll(final Long boardId) {
        log.info("Method=findAllByBoardId, boardId={}", boardId);
        List<LeadTimeConfig> leadTimeConfigs = leadTimeConfigRepository.findByBoardId(boardId);

        return leadTimeConfigMapper.toResponse(leadTimeConfigs);
    }

    @Override
    @Transactional
    public Long create(final Long boardId, final LeadTimeConfigRequest leadTimeConfigRequest) {
        log.info("Method=create, boardId={}, leadTimeConfigRequest={}", boardId, leadTimeConfigRequest);

        Board board = boardService.findById(boardId);
        LeadTimeConfig leadTimeConfig = leadTimeConfigMapper.toLeadTimeConfig(leadTimeConfigRequest, board.getId());

        leadTimeConfigRepository.save(leadTimeConfig);

        return leadTimeConfig.getId();
    }

    @Override
    @Transactional(readOnly = true)
    public LeadTimeConfigResponse findByBoardAndId(final Long boardId, final Long id) {
        log.info("Method=findByBoardAndId, boardId={}, id={}", boardId, id);

        LeadTimeConfig leadTimeConfig = leadTimeConfigRepository.findByIdAndBoardId(id, boardId)
                .orElseThrow(ResourceNotFound::new);
        return leadTimeConfigMapper.toResponse(leadTimeConfig);
    }

    @Override
    @Transactional
    public void update(final Long boardId, final Long id, final LeadTimeConfigRequest leadTimeConfigRequest) {
        log.info("Method=update, boardId={}, id={}, leadTimeConfigRequest={}", boardId, id, leadTimeConfigRequest);

        LeadTimeConfig leadTimeConfig = leadTimeConfigRepository.findByIdAndBoardId(id, boardId)
                .orElseThrow(ResourceNotFound::new);

        leadTimeConfigMapper.updateFromRequest(leadTimeConfig, leadTimeConfigRequest);
        leadTimeConfigRepository.save(leadTimeConfig);
    }

    @Override
    @Transactional
    public void deleteByBoardAndId(final Long boardId, final Long id) {
        log.info("Method=deleteByBoardAndId, boardId={}, id={}", boardId, id);

        LeadTimeConfig leadTimeConfig = leadTimeConfigRepository.findByIdAndBoardId(id, boardId)
                .orElseThrow(ResourceNotFound::new);
        leadTimeConfigRepository.delete(leadTimeConfig);
    }
}

File: src/main/java/br/com/jiratorio/service/impl/LeadTimeConfigServiceImpl.kt
package br.com.jiratorio.service.impl

import br.com.jiratorio.domain.entity.LeadTimeConfig
import br.com.jiratorio.domain.request.LeadTimeConfigRequest
import br.com.jiratorio.domain.response.LeadTimeConfigResponse
import br.com.jiratorio.exception.ResourceNotFound
import br.com.jiratorio.mapper.LeadTimeConfigMapper
import br.com.jiratorio.repository.LeadTimeConfigRepository
import br.com.jiratorio.service.BoardService
import br.com.jiratorio.service.LeadTimeConfigService
import br.com.jiratorio.extension.logger
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class LeadTimeConfigServiceImpl(
    private val leadTimeConfigRepository: LeadTimeConfigRepository,
    private val boardService: BoardService,
    private val leadTimeConfigMapper: LeadTimeConfigMapper
) : LeadTimeConfigService {

    private val log = logger()

    @Transactional(readOnly = true)
    override fun findAllByBoardId(boardId: Long): List<LeadTimeConfig> {
        log.info("Method=findAllByBoardId, boardId={}", boardId)

        return leadTimeConfigRepository.findByBoardId(boardId)
    }

    @Transactional(readOnly = true)
    override fun findAll(boardId: Long): List<LeadTimeConfigResponse> {
        log.info("Method=findAllByBoardId, boardId={}", boardId)

        val leadTimeConfigs = leadTimeConfigRepository.findByBoardId(boardId)
        return leadTimeConfigMapper.toResponse(leadTimeConfigs)
    }

    @Transactional
    override fun create(boardId: Long, leadTimeConfigRequest: LeadTimeConfigRequest): Long {
        log.info("Method=create, boardId={}, leadTimeConfigRequest={}", boardId, leadTimeConfigRequest)

        val board = boardService.findById(boardId)
        val leadTimeConfig = leadTimeConfigMapper.toLeadTimeConfig(leadTimeConfigRequest, board)

        leadTimeConfigRepository.save(leadTimeConfig)

        return leadTimeConfig.id
    }

    @Transactional(readOnly = true)
    override fun findByBoardAndId(boardId: Long, id: Long): LeadTimeConfigResponse {
        log.info("Method=findByBoardAndId, boardId={}, id={}", boardId, id)

        val leadTimeConfig = leadTimeConfigRepository.findByIdAndBoardId(id, boardId)
            .orElseThrow(::ResourceNotFound)

        return leadTimeConfigMapper.toResponse(leadTimeConfig)
    }

    @Transactional
    override fun update(boardId: Long, id: Long, leadTimeConfigRequest: LeadTimeConfigRequest) {
        log.info("Method=update, boardId={}, id={}, leadTimeConfigRequest={}", boardId, id, leadTimeConfigRequest)

        val leadTimeConfig = leadTimeConfigRepository.findByIdAndBoardId(id, boardId)
            .orElseThrow(::ResourceNotFound)

        leadTimeConfigMapper.updateFromRequest(leadTimeConfig, leadTimeConfigRequest)
        leadTimeConfigRepository.save(leadTimeConfig)
    }

    @Transactional
    override fun deleteByBoardAndId(boardId: Long, id: Long) {
        log.info("Method=deleteByBoardAndId, boardId={}, id={}", boardId, id)

        val leadTimeConfig = leadTimeConfigRepository.findByIdAndBoardId(id, boardId)
            .orElseThrow(::ResourceNotFound)

        leadTimeConfigRepository.delete(leadTimeConfig)
    }
}
